File: aingle-back/src/main/java/com/aintopia/aingle/alarm/domain/Alarm.java
package com.aintopia.aingle.alarm.domain;

import com.aintopia.aingle.member.domain.Member;
import com.aintopia.aingle.post.domain.Post;
import com.aintopia.aingle.vote.domain.Vote;
import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.ColumnDefault;
import org.hibernate.annotations.DynamicInsert;
import org.hibernate.annotations.DynamicUpdate;

import java.time.LocalDateTime;

@Entity
@Table(name = "alarm")
@DynamicInsert
@DynamicUpdate
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Getter
public class Alarm {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "alarm_id")
    private Long alarmId;

    @Column(name = "is_read")
    @ColumnDefault("false")
    private Boolean isRead;

    @Column(name = "create_time", columnDefinition = "TIMESTAMP")
    private LocalDateTime createTime;

    @ManyToOne(optional = false, fetch = FetchType.EAGER)
    @JoinColumn(name = "member_id")
    private Member member;

    @ManyToOne(optional = false, fetch = FetchType.EAGER)
    @JoinColumn(name = "sender_id")
    private Member sender;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "post_id")
    private Post post;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "vote_id")
    private Vote vote;

    @Builder(builderMethodName = "alarmPostBuilder")
    public Alarm(Post post, Member member, Member sender) {
        this.post = post;
        this.member = member;
        this.sender = sender;
    }


    public void isRead() {
        this.isRead = true;
    }
}

File: aingle-front/src/model/alarm.ts
import { IvoteWinnerCharacter } from "./vote";

// alarm에 포함되는 알림 sender 정보
export interface Isender {
    memberId : number;
    name : string;
    memberImage : string;
}

// alarm에 포함되는 게시글 정보
export interface Ipost {
    postId : number;
    image : string;
}

export interface Ialarm {
  alarmId: number;
  isRead: boolean;
  // vote 타입의 알림인 경우 post가 없기 때문에 null 처리
  post: Ipost | null;
  // 투표 알림이 아닐 경우 null이 반환되기 때문에 null 처리
  votewinnerCharacter: IvoteWinnerCharacter | null;
  createTime: string;
  // vote 타입의 알림인 경우 sender가 없기 때문에 null 처리
  sender: Isender | null;
}

export interface IgetAlarm {
    page : number;
    size : number;
}